### PR TITLE
feat(git/auth): Support hostrules with username and password when authenticating git commands

### PR DESCRIPTION
**Pull request by Shegox**
## Changes

This pull request introduces the ability to authenticate git commands using a combination of username and password through host rules in addition to the currently supported token format.

## Context

Fixes #24077

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/git/auth.ts [hunk 1]
```diff
@@ -59,11 +59,9 @@ export function getGitAuthenticatedEnvironmentVariables(
...
   }
 
-  const authenticationRules = getAuthenticationRulesWithToken(
-    originalGitUrl,
-    hostType,
-    token
-  );
+  const authenticationRules = token
+    ? getAuthenticationRulesWithToken(originalGitUrl, hostType, token)
+    : getAuthenticationRules(originalGitUrl, `${username}:${password}`);
```
Comment (Shegox) [inline lib/util/git/auth.ts:64]: One question I had during implementation is if we should/need to URL encode the username/password here?

The token is not encoded, but the chance that some special URL characters like `/` `@` are part of the password is way more likely then in a token.

File: lib/util/git/auth.ts [hunk 2]
```diff
@@ -151,15 +149,18 @@ export function getGitEnvironmentVariables(
...
-  const githubToken = find({
+  const gitHubHostRule = find({
     hostType: 'github',
     url: 'https://api.github.com/',
   });
 
-  if (githubToken?.token) {
+  if (
+    gitHubHostRule?.token ??
+    (gitHubHostRule?.username && gitHubHostRule?.password)
```
Comment (viceice) [inline lib/util/git/auth.ts:159]: revert, username and password doesn't work for github.com

File: lib/util/git/auth.ts (hunk 1 above)
Comment (viceice) [reply lib/util/git/auth.ts:64]: probably yes, we need to add to sanitize after encoding too

Comment (Shegox) [reply lib/util/git/auth.ts]: Added the encoding. 

While testing it I noticed that they are already getting correctly sanitized from the logs and I have no idea  yet why 😅

Comment (Shegox) [reply lib/util/git/auth.ts]: Found it: 

https://github.com/renovatebot/renovate/blob/53263bbfda3eebf20506defd8ae78b9c53048766/lib/logger/utils.ts#L320-L324

Already removes all the username+password things from URLs

File: lib/modules/manager/gomod/readme.md [hunk 3]
```diff
@@ -32,10 +32,10 @@ Before running the `go` commands to update the `go.sum`, Renovate exports `git`
 
 The following logic is executed prior to "artifacts" updating:
 
-The token from the `hostRules` entry matching `hostType=github` and `matchHost=api.github.com` is added as the default authentication for `github.com`.
+The token or username/password from the `hostRules` entry matching `hostType=github` and `matchHost=api.github.com` is added as the default authentication for `github.com`.
```
Comment (viceice) [inline lib/modules/manager/gomod/readme.md:35]: same

File: lib/util/git/auth.spec.ts [hunk 4]
```diff
@@ -28,6 +28,45 @@ describe('util/git/auth', () => {
       });
     });
 
+    it('returns url with username and password', () => {
+      expect(
+        getGitAuthenticatedEnvironmentVariables('https://github.com/', {
```
Comment (viceice) [inline lib/util/git/auth.spec.ts:33]: don't do that for GitHub.com. password auth is not working there. it was disabled long time ago.

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/util/git/auth.spec.ts (hunk 4 above)
Comment (Shegox) [reply lib/util/git/auth.spec.ts:33]: username:password auth does work for github.com/GitHub Enterprise Server as long as the password is an access token. So you can technically supply Renovate with a `usernane: some-user`, `password: ghp_token` and it will work

File: lib/util/git/auth.ts [hunk 5]
```diff
@@ -151,15 +161,15 @@ export function getGitEnvironmentVariables(
   let environmentVariables: NodeJS.ProcessEnv = {};
 
   // hard-coded logic to use authentication for github.com based on the githubToken for api.github.com
-  const githubToken = find({
+  const gitHubHostRule = find({
```
Comment (viceice) [inline lib/util/git/auth.ts:164]: Is this rename required?

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/util/git/auth.ts (hunk 5 above)
Comment (Shegox) [reply lib/util/git/auth.ts:164]: It isn't required, but while looking through it I noticed that `gitHubHostRule` reflects the purpose better than `githubToken`.

Comment (Shegox) [reply lib/util/git/auth.ts:164]: should I move this change in a separate PR?

**Review by viceice — APPROVED** [review]

**Comment by lcobucci** [comment]
Just tested the fix and I still get the authentication errors :cry:
I'll do some local debugging and come back you.

Thanks @Shegox and @viceice for working on this :+1:

**Comment by lcobucci** [comment]
Okay, after adding a whole lot of `console.log()` on a few places I've managed to find the issue... it's a tricky one, I think :sob: 

We were previously relying on the host types `git-refs` and `git-tags` to resolve the submodules.
The authentication changes aren't considering those host types, leading the creation of environment variables using the token as auth mechanism (`RENOVATE_TOKEN` is automatically promoted as a host rule).

So, essentially I have the following host rules:

```js
[
  {
    hostType: 'github',
    matchHost: 'github.com',
    token: 'github-token',
    resolvedHost: 'github.com'
  },
  {
    hostType: 'git-refs',
    matchHost: 'mygitlab.com',
    password: 'mygitlab-pwd',
    username: 'renovate-bot',
    resolvedHost: 'mygitlab.com'
  },
  {
    hostType: 'git-tags',
    matchHost: 'mygitlab.com',
    password: 'mygitlab-pwd',
    username: 'renovate-bot',
    resolvedHost: 'mygitlab.com'
  },
  {
    matchHost: 'mygitlab.com',
    token: 'mygitlab-pwd',
    hostType: 'gitlab',
    resolvedHost: 'mygitlab.com'
  }
]
```

The host rule created from `RENOVATE_TOKEN` is legit for the API requests. However, not for the Git related requests.

Filtering for `git-refs`, `renovate` creates the following env vars (contents of `newEnvironmentVariables`):

```js
{
  GIT_CONFIG_KEY_0: 'url.https://ssh:github-token@github.com/.insteadOf',
  GIT_CONFIG_VALUE_0: 'ssh://git@github.com/',
  GIT_CONFIG_KEY_1: 'url.https://git:github-token@github.com/.insteadOf',
  GIT_CONFIG_VALUE_1: 'git@github.com:',
  GIT_CONFIG_KEY_2: 'url.https://github-token@github.com/.insteadOf',
  GIT_CONFIG_VALUE_2: 'https://github.com/',
  GIT_CONFIG_COUNT: '6',
  GIT_CONFIG_KEY_3: 'url.https://renovate-bot:mygitlab-pwd@mygitlab.com/.insteadOf',
  GIT_CONFIG_VALUE_3: 'ssh://git@mygitlab.com/',
  GIT_CONFIG_KEY_4: 'url.https://renovate-bot:mygitlab-pwd@mygitlab.com/.insteadOf',
  GIT_CONFIG_VALUE_4: 'git@mygitlab.com:',
  GIT_CONFIG_KEY_5: 'url.https://renovate-bot:mygitlab-pwd@mygitlab.com/.insteadOf',
  GIT_CONFIG_VALUE_5: 'https://mygitlab.com/'
}
```
And manually running `git ls-remote --symref https://mygitlab.com/path/to/my/repo.git HEAD` with those variables work just fine. However, letting the command run leads to the password prompt of a completely weird username (`oauth2`) - which understandably fails on non-interactive sessions.

I also confirm that running `git ls-remote --symref https://oauth2@mygitlab.com/path/to/my/repo.git HEAD` with those variables **doesn't work at all** (which makes sense because there's no match on `https://oauth2@...`.

What do you think? Are we might still be missing some scenarios for `getAuthenticationRules()` or `getGitEnvironmentVariables()`...

**Comment by lcobucci** [comment]
> please open a new discussion

@viceice the original hasn't been concluded IMHO so I just added context [here](https://github.com/renovatebot/renovate/discussions/24077#discussioncomment-6906131). I believe it's quite counter-productive to open a new one.

**Comment by viceice** [comment]
you can of cause use the existing discussion, but not the PR. this PR is wrong place to discuss things.
